CORE-2950 schema_registry: improve authn/authz audit logging

pgellert opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Improve the audit logging messages for authentication and authorization failures in schema registry.

Fixes https://redpandadata.atlassian.net/browse/CORE-2950

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/49150#018f7bfd-5e71-4634-8c26-1473e3b247f4:
```
"rptest.tests.audit_log_test.AuditLogTestSchemaRegistry.test_sr_audit_bad_authn"
```
<!--04f6c36f-2e9e-496f-bd43-8c66e6a020bf-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/49150#018f7c05-8931-48ce-9d19-f92261006442
<!--d2ace6a7-4c2b-4e31-8223-5b92f2640d90-->

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/b8bcfcc3c4daabe2e31044438931796029fbafce..2b24ab93d23ff9ec2ad553c9c01e387c0abe759b) with various changes, including:
* Fixed the above failing ducktape test
* Added a ducktape test to exercise authz failures and its audit logging
* Made various fixes to auth/audit handling and refactoring in ducktape to achieve the above

oleiman commented:
@pgellert - Appreciate the effort to break this down into bite-sized commits. [the first chunk](https://github.com/redpanda-data/redpanda/compare/209f78d513b5c5923b51a02b7dceb841d9ed24b1..75915a7e3a0556fb877df3cdcf80e8f106dccb67) might be easier to grok if consolidated into 2 commits:
- one relocating all of the code from auth_utils.h -> service.cc w/o changing any functionality
- and another adding the required auditing stuff

wdyt?

pgellert commented:
@oleiman yeah that makes sense. I've squashed these into 2 commits now. (noop [force-push](https://github.com/redpanda-data/redpanda/compare/2df78261f321eaaeea46ba29b9b9c9e3d7db3c1d..e3d56eaeb570fc3c45e33e121edf1017f9eac4eb))

dotnwat commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -100,6 +90,78 @@ class wrap {
+                auth_result.pass();
+                break;
+            }
+        } catch (unauthorized_user_exception& e) {
```
`const&`

dotnwat commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -55,6 +54,15 @@ const security::acl_principal principal{
 
 class wrap {
 public:
+    enum class auth_level {
+        // Unauthenticated endpoint (not a typo, 'public' is a keyword)
+        publik = 0,
+        // Requires authentication (if enabled) but not superuser status
+        user = 1,
+        // Requires authentication (if enabled) and superuser status
+        superuser = 2
+    };
```
if the enumerator values (e.g. 0,1,2) aren't important, it's better to remove them (e.g. their values neither enter nor escape redpanda process). if they are important, a note about where their values interact with the outside world is a useful comment to have.

oleiman reviewed (commented):
looks legit. still need to go over the tests

oleiman commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1924,84 +1993,44 @@ def test_sr_audit(self):
         _ = self.find_matching_record(
-            lambda record: match_authn_user(self.username, self.
-                                            sr_audit_svc_name, 1, record),
+            lambda record: self.match_authn_record(record, StatusID.SUCCESS),
```
question (not related to your change): any idea why we don't assert here? like what is this testing?

pgellert replied:
The "assertion" here is implicit because `find_matching_record` would throw a `TimeoutError` if we don't find a matching record.

oleiman commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1924,84 +1993,44 @@ def test_sr_audit(self):
-            assert f"Should not have found any records but found {self.aggregate_count(records)}: {records}"
+            assert False, f"Should not have found any records but found {self.aggregate_count(records)}: {records}"
         except TimeoutError:
             pass
```
suggestion(non-blocking) - `expect_exception(TimeoutError, lambda _: True)` should work here.

pgellert replied:
Yeah, that's much nicer, thanks

oleiman commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1963,11 +1963,17 @@ def match_authn_record(self, record, status_id: StatusID):
             record['dst_endpoint']['svc_name'] == self.sr_audit_svc_name:
             self.logger.debug(f"Validating api activity record: {record}")
 
+        if status_id and record.get('status_id', '') == str(status_id.value):
+            return False
+
```
should this be `!=`?

pgellert replied:
Yeah, good spot, thanks. I've fixed this now.

pgellert commented:
[Force-pushed](https://github.com/redpanda-data/redpanda/compare/e3d56eaeb570fc3c45e33e121edf1017f9eac4eb..b93ac38bc185f64934d9910ba8e9b4101907c13b) to address the review feedback above.

Also, I fixed the failing ducktape test by waiting for the user creation to be observable, as otherwise, we could sometimes have auth failures before the user was visible even though we expected auth to succeed because we have called `create_user`.

oleiman reviewed (approved):
lgtm!

BenPope commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -131,14 +117,49 @@ class wrap {
 
-        return user;
+            // Log successful authn
+            audit_authn(rq);
```
I wonder if this could be reworked so that auditing success and failure are not in different functions? Would make it easier to eyeball if all the cases are covered.

pgellert replied:
Extracted this code to a lambda now

BenPope commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -194,6 +215,31 @@ class wrap {
+        if (!success) {
+            vlog(
+              plog.error,
+              "Failed to audit authorization request for endpoint: {}",
+              rq.req->format_url());
```
I don't know if this is a new bug, but I'm seeing the query params repeated.

Same for authn (and it has a typo: "Failed to audit authnetication request for endpoint: {}"

pgellert replied:
I fixed the typo, thanks.

The bug existed before as well. I had a quick look but I couldn't find where it's coming from just yet.

BenPope replied:
I think `format_url` :'(

pgellert commented:
I've moved all the code into `handle_auth`: https://github.com/redpanda-data/redpanda/compare/1b2fc0f797ca4f5522ab8bc40f931945934484c6..88b555a41f1c11b670d6f5c0430e659bdae23c17 to address some more code review feedback.

pgellert commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -108,19 +90,23 @@ class wrap {
-      const ss::http::request& req) const {
-        credential_t user;
-
-        if (authn_method != config::rest_authn_method::none) {
```
Extracted this code to a lambda now

pgellert commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -108,19 +90,23 @@ class wrap {
-            // did not give the authorization header.
-            switch (lvl) {
+    request_auth_result
+    maybe_authenticate_request(const server::request_t& rq) const {
```
Extracted this code to a lambda now

pgellert commented:
[Fixed](https://github.com/redpanda-data/redpanda/compare/88b555a41f1c11b670d6f5c0430e659bdae23c17..8b5517a50d46026abed36ed1eeafe6927f20fc58) the test doc comment.

BenPope commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1911,9 +1953,52 @@ def __init__(self, test_context):
             schema_registry_config=sr_config)
 
+    def match_authn_record(self, record, status_id: StatusID):
+        if record['class_uid'] == ClassUID.AUTHENTICATION and record[
+                'dst_endpoint']['svc_name'] == self.sr_audit_svc_name:
+            self.logger.debug(f"Validating auth record: {record}")
+
+        return record['class_uid'] == ClassUID.AUTHENTICATION and \
+            record['dst_endpoint']['svc_name'] == self.sr_audit_svc_name and \
+            record['user']['name'] == self.username and \
+            record['status_id'] == status_id
```
nitpick: I think it's clearer with early-exit:
```suggestion
        if record['class_uid'] != ClassUID.AUTHENTICATION or record[
                'dst_endpoint']['svc_name'] != self.sr_audit_svc_name:
            return False

        self.logger.debug(f"Validating auth record: {record}")
        return record['user']['name'] == self.username and record[
            'status_id'] == status_id

BenPope commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1911,9 +1953,52 @@ def __init__(self, test_context):
+                         record,
+                         endpoint,
+                         status_id: Optional[StatusID] = None):
+        if record['class_uid'] == ClassUID.API_ACTIVITY and \
+            record['dst_endpoint']['svc_name'] == self.sr_audit_svc_name:
+            self.logger.debug(f"Validating api activity record: {record}")
+
+        if status_id and record.get('status_id', '') != status_id:
+            return False
+
+        if record['class_uid'] == ClassUID.API_ACTIVITY \
+            and record['dst_endpoint']['svc_name'] == self.sr_audit_svc_name \
+            and record['actor']['user']['name'] == self.username:
+            regex = re.compile(
+                "http:\/\/(?P<address>.*):(?P<port>\d+)\/(?P<handler>.*)")
+            url_string = record['http_request']['url']['url_string']
+            match = regex.match(url_string)
+            if match and match.group('handler') == endpoint:
+                return True
+
+        return False
```
nitpick: I think it's clearer with early exit (I also took the liberty of using a raw string for the regex):
```suggestion
        if record['class_uid'] != ClassUID.API_ACTIVITY or record[
                'dst_endpoint']['svc_name'] != self.sr_audit_svc_name:
            return False

        self.logger.debug(f"Validating api activity record: {record}")
        if (status_id and record.get('status_id', '') != status_id
            ) or record['actor']['user']['name'] != self.username:
            return False

        regex = re.compile(
            r"http://(?P<address>.*):(?P<port>\d+)/(?P<handler>.*)")
        url_string = record['http_request']['url']['url_string']
        match = regex.match(url_string)
        return match and match.group('handler') == endpoint
```

BenPope commented on tests/rptest/tests/audit_log_test.py:
```diff
@@ -1911,9 +1953,52 @@ def __init__(self, test_context):
         self.admin.create_user(self.username, self.password, self.algorithm)
 
+        # wait for user to propagate to nodes
+        def user_exists():
+            for node in self.redpanda.nodes:
+                users = self.admin.list_users(node=node)
+                if self.username not in users:
+                    return False
+            return True
```
nitpick:
```suggestion
            return all(self.username in self.admin.list_users(node=node)
                       for node in self.redpanda.nodes)
```

michael-redpanda commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -108,19 +96,31 @@ class wrap {
+        try {
+            return rq.service().authenticator().authenticate(*rq.req);
+        } catch (unauthorized_user_exception& e) {
+            audit_authn_failure(rq, "", e.what());
```
n/m old comment

dotnwat commented on src/v/pandaproxy/schema_registry/service.cc:
```diff
@@ -100,6 +90,62 @@ class wrap {
+
+            // Will throw 403 if user enabled HTTP Basic Auth but
+            // did not give the authorization header.
+            [this, &rq, &auth_result]() {
+                try {
+                    switch (_auth_level) {
+                    case auth_level::superuser:
+                        auth_result.require_superuser();
+                        break;
+                    case auth_level::user:
+                        auth_result.require_authenticated();
+                        break;
+                    case auth_level::publik:
+                        auth_result.pass();
+                        break;
+                    }
+                } catch (const ss::httpd::base_exception& e) {
+                    audit_authz_failure(rq, auth_result, e.what());
+                    throw;
+                }
+            }();
```
why is this an IIL? it's not being used to initialize a variable, which is the most common use case.
